### ISSUE TITLE
`cls()` vs `cls.__call__()` in class methods with custom metaclass

### ISSUE THREAD

**Issue by fill1890** [microsoft/pyright#5869]
Hi, I'm not sure if this expected behaviour given the complexity of class instantiation, but if I use a classmethod in a class with a custom metaclass pyright seems to report the call signature incorrectly.

Can be reproduced as follows:
```python
class Other:
    pass

class Meta(type):
    def __call__(cls, *args, **kwargs) -> Other:
        return Other()

class Test(metaclass=Meta):
    @classmethod
    def test(cls) -> Other:
        reveal_type(cls())             # Type of "cls()" is "Self@Base"
        reveal_type(cls.__call__())    # Type of "cls.__call__()" is "Other"
        return cls()                   # Expression of type "Self@Base" cannot be assigned to "Other"
```
I would assume the signature of the metaclass `__call__` method should be used for `cls()`?
Thanks :)

Pyright 1.1.325 as a neovim LSP and on the command line

**Comment by erictraut** [comment]
Yes, this is expected behavior. The first parameter for a class method is inferred to be `type[Self]`. 

What you're doing here is highly unusual and not compatible with static typing. It breaks assumptions that static type checkers must make.

My recommendation is to avoid this technique if you are interested in having your code work well with static typing. If that's not an option, you can use an explicit type annotation for the `cls` parameter and suppress the resulting type error.

```python
class Test(metaclass=Meta):
    @classmethod
    def test(cls: type[Other]) -> Other: # type: ignore
        reveal_type(cls())
        reveal_type(cls.__call__())
        return cls()
```
